Repurpose Gen0ReductionCount field for dynamic heap count diagnostics.

PeterSolMS opened the pull request:
A quick hack to make diagnostics for dynamic heap count easier.

PeterSolMS committed:
Repurpose Gen0ReductionCount field for dynamic heap count diagnostics.

PeterSolMS committed:
Divide msl channels by the heap count to give a more accurate impression of the impact.

PeterSolMS committed:
Improvements and bug fixes:
- subtract  time spent in safe_switch_to_thread and WaitLongerNoInstru this reduces the noise in the msl overhead channel
- refactor enter_spin_lock_msl so the fast path can be inlined and the slow path is in a separate helper
- additional instrumentation in check_heap_count and change_heap_count
- fix hang bug in change_heap_count -  issue was that we were comparing against n_heaps which may already have changed
- add spinning to EnterFinalizeLock

Maoni0 commented:
sorry I wasn't being clear - I never meant that we were going to check in the repurpose part - it was just there for testing purposes. so we wouldn't need to wait for the parsing work to be done _to look at the traces_. the parsing work isn't high priority for testing since we could easily find some fields to repurpose.

the rest of this PR is obviously needed and should be included in the build for testing purposes. @PeterSolMS do you mind creating another PR "adaptation fixes and improvements" that encapsulates the hang fix and the msl overhead instability/finalizer lock improvement?

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -1413,7 +1413,7 @@ bool gc_heap::should_move_heap (GCSpinLock* msl)
 // All the places where we could be stopped because there was a suspension should call should_move_heap to check if we need to return
 // so we can try another heap or we can continue the allocation on the same heap.
-enter_msl_status gc_heap::enter_spin_lock_msl (GCSpinLock* msl)
+enter_msl_status gc_heap::enter_spin_lock_msl_helper (GCSpinLock* msl)
```
can we change this method a bit so we don't immediately need to do another Interlocked check? you could change it to a do/while loop. 

also we should run some microbenchmarks to see if this regresses anything (especially WKS GC)...

PeterSolMS committed:
Adress code review feedback - change enter_spin_lock_msl_helper to use do-while loop to avoid doing an immediate lock compare exchange after the failing one.

Change reporting of dynamic heap count values to report to 0.1% precision, drop #allocating threads.

Add reporting of median overhead, percentage point overhead reduction/increase and space cost reduction/increase to 0.1% precision.

PeterSolMS committed:
Fix stale comment

PeterSolMS committed:
Merge branch 'main' into Repurpose_ETW_Event

Maoni0 reviewed (approved):
LGTM!

Maoni0 commented:
Actually, talking to Moko just now made me realize that it’s probably better if we make it perfectly obviously if adaption code is there but adaptation is not turned on so maybe in that case we just make the repurposed builds have -1 for those fields?

PeterSolMS committed:
Address code review feedback - if dynamic heap count is not enabled because of configs, set the two repurposed fields to -1 (all 1 bits).

Maoni0 commented:
hmm, I'm not sure what happened? I wanted to push some more (temporary) instrumentation changes to this PR but it instantly triggered some sort wide code review requests... sorry about that! please ignore!

Commit:
repurpose the CreateSegment event for commit info

mangod9 commented:
@cshung @mrsharm assume we could close this since we plan to enable dynamic events?

mrsharm commented:
This PR consists of instrumentation and fixes. The fixes have been merged with the following PR: https://github.com/dotnet/runtime/pull/87218/files - we should close this PR and extract the instrumentation logic.
